pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpartition/SegmentPartitionMetadataManager.java

```diff
@@ -123,16 +141,23 @@ private List<String> getOnlineServers(ExternalView externalView, String segment)
 
   private void computeTablePartitionInfo() {
     PartitionInfo[] partitionInfoMap = new PartitionInfo[_numPartitions];
-    Set<String> segmentsWithInvalidPartition = new HashSet<>();
+    List<String> segmentsWithInvalidPartition = new ArrayList<>();
+    List<Map.Entry<String, SegmentInfo>> newSegmentInfoEntries = new ArrayList<>();
+    long currentTimeMs = System.currentTimeMillis();
     for (Map.Entry<String, SegmentInfo> entry : _segmentInfoMap.entrySet()) {
       String segment = entry.getKey();
       SegmentInfo segmentInfo = entry.getValue();
       int partitionId = segmentInfo._partitionId;
-      List<String> onlineServers = segmentInfo._onlineServers;
       if (partitionId == INVALID_PARTITION_ID) {
         segmentsWithInvalidPartition.add(segment);
         continue;
       }
+      // Process new segments in the end
+      if (InstanceSelector.isNewSegment(segmentInfo._pushTimeMs, currentTimeMs)) {
```

Comment:
is it possible that a new segment is with invalid partitionID? should we move this above the invalid id check?

Reply:
New segment should have valid partition id. Segment ZK metadata won't be changed for new segment within the life cycle